[FLINK-31238] Use IngestDB to speed up Rocksdb rescaling recovery (rebased)

StefanRRichter opened the pull request:
Rebased and slightly refactored version of #23169.

StefanRRichter commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/restore/RocksDBIncrementalRestoreOperation.java:
```diff
@@ -414,90 +424,70 @@ private void restoreWithRescaling(Collection<KeyedStateHandle> restoreStateHandl
+                    for (Pair<RegisteredStateMetaInfoBase, ExportImportFilesMetaData> entry :
+                            exportedCFAndMetaData) {
+                        ExportImportFilesMetaData cfMetaData = entry.getValue();
+                        // TODO: method files() doesn't exist in the RocksDB API
```
@mayuehappy Is it correct to remove this code? I could not find a files() method in the RocksDB API.

rkhachatryan commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/EmbeddedRocksDBStateBackend.java:
```diff
@@ -305,6 +311,11 @@ private EmbeddedRocksDBStateBackend(
                 overlapFractionThreshold >= 0 && this.overlapFractionThreshold <= 1,
                 "Overlap fraction threshold of restoring should be between 0 and 1");
 
+        useIngestDbRestoreMode =
+                original.useIngestDbRestoreMode == UNDEFINED_USE_INGEST_DB_RESTORE_MODE
+                        ? config.get(USE_INGEST_DB_RESTORE_MODE)
+                        : original.useIngestDbRestoreMode;
```
Won't this ignore `original.useIngestDbRestoreMode` if it's set to `false`?
(I guess this is a copy-paste from other fields, but they aren't boolean)

StefanRRichter replied:
Yep, I think we better use `Boolean`and `null` for undefined.

rkhachatryan commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBIncrementalCheckpointUtils.java:
```diff
@@ -154,6 +162,54 @@ private static void deleteRange(
+    public static List<Pair<RegisteredStateMetaInfoBase, ExportImportFilesMetaData>>
+            exportColumnFamilies(
+                    RocksDB db,
+                    List<ColumnFamilyHandle> columnFamilyHandles,
+                    List<StateMetaInfoSnapshot> stateMetaInfoSnapshots,
```
This to lists should match each other, right?
Should we check with `Preconditions.checkState(columnFamilyHandles.size() == stateMetaInfoSnapshots.size());`,
or better pass RocksDB instance here?

rkhachatryan commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBOperationUtils.java:
```diff
@@ -230,6 +252,19 @@ private static ColumnFamilyHandle createColumnFamily(
+        try {
+            return db.createColumnFamilyWithImport(
+                    columnDescriptor, new ImportColumnFamilyOptions(), metaDataList);
+        } catch (RocksDBException e) {
+            IOUtils.closeQuietly(columnDescriptor.getOptions());
+            throw new FlinkRuntimeException("Error creating ColumnFamilyHandle.", e);
```
Missing `metaDataList.forEach(IOUtils::closeQuietly);` ?

StefanRRichter replied:
I modified the code in a different way that makes sure the cleanup happens through the owner of the objects.

rkhachatryan commented on flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksIncrementalCheckpointRescalingTest.java:
```diff
@@ -419,7 +431,12 @@ private KeyedOneInputStreamOperatorTestHarness<String, String, Integer> getHarne
-        return new RocksDBStateBackend("file://" + rootFolder.newFolder().getAbsolutePath(), true);
+        RocksDBStateBackend rocksDBStateBackend =
+                new RocksDBStateBackend("file://" + rootFolder.newFolder().getAbsolutePath(), true);
+        Configuration configuration = new Configuration();
```
Missing import.

rkhachatryan commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/restore/RocksDBIncrementalRestoreOperation.java:
```diff
@@ -329,74 +332,164 @@ private void restoreWithRescaling(Collection<KeyedStateHandle> restoreStateHandl
+            columnFamilyMetaDataToImport.forEach(
+                    this.rocksHandle::registerStateColumnFamilyHandleWithImport);
+            logger.info(
+                    "Finished importing exported column families into target DB for rescaling.");
         } finally {
-            // Cleanup all download directories
-            allDownloadSpecs.stream()
-                    .map(StateHandleDownloadSpec::getDownloadDestination)
-                    .forEach(this::cleanUpPathQuietly);
+            cleanUpPathQuietly(exportCfBasePath);
```
Just to clarify, `cleanUpPathQuietly(exportCfBasePath)` doesn't affect the opened RocksDb instance by deleting the imported files?

StefanRRichter replied:
No, I think the files are hard-linked in the directory of the created instance to which we import.

rkhachatryan commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/restore/RocksDBIncrementalRestoreOperation.java:
```diff
@@ -329,74 +332,164 @@ private void restoreWithRescaling(Collection<KeyedStateHandle> restoreStateHandl
+            throws Exception {
+
+        final Path absolutInstanceBasePath = instanceBasePath.getAbsoluteFile().toPath();
+        final Path exportCfBasePath = absolutInstanceBasePath.resolve("export-cfs");
```
Just to clarify, this is not a temporary folder so that the files can be efficiently sym-linked?

rkhachatryan commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/EmbeddedRocksDBStateBackend.java:
```diff
@@ -179,6 +182,8 @@ public class EmbeddedRocksDBStateBackend extends AbstractManagedMemoryStateBacke
      */
     private double overlapFractionThreshold;
 
+    private boolean useIngestDbRestoreMode;
+
```
Can be `final` according to IntelliJ :)

rkhachatryan commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/restore/RocksDBHandle.java:
```diff
@@ -185,6 +187,33 @@ RocksDbKvStateInfo getOrRegisterStateColumnFamilyHandle(
+            RegisteredStateMetaInfoBase stateMetaInfo,
+            List<ExportImportFilesMetaData> cfMetaDataList) {
+
+        RocksDbKvStateInfo registeredStateMetaInfoEntry =
+                kvStateInformation.get(stateMetaInfo.getName());
+
+        Preconditions.checkState(registeredStateMetaInfoEntry == null);
```
Replace with
```
Preconditions.checkState(!kvStateInformation.containsKey(stateMetaInfo.getName()));
``` 
?

rkhachatryan commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/restore/RocksDBHandle.java:
```diff
@@ -185,6 +187,33 @@ RocksDbKvStateInfo getOrRegisterStateColumnFamilyHandle(
+
+        Preconditions.checkState(registeredStateMetaInfoEntry == null);
+
+        registeredStateMetaInfoEntry =
+                RocksDBOperationUtils.createStateInfo(
+                        stateMetaInfo,
+                        db,
+                        columnFamilyOptionsFactory,
+                        ttlCompactFiltersManager,
+                        writeBufferManagerCapacity,
+                        cfMetaDataList);
+
+        RocksDBOperationUtils.registerKvStateInformation(
+                kvStateInformation,
+                nativeMetricMonitor,
+                stateMetaInfo.getName(),
+                registeredStateMetaInfoEntry);
+
+        return registeredStateMetaInfoEntry;
```
The return value isn't used, so return `void` and inline?

rkhachatryan commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/restore/RocksDBIncrementalRestoreOperation.java:
```diff
@@ -105,6 +110,10 @@ public class RocksDBIncrementalRestoreOperation<K> implements RocksDBRestoreOper
 
     private boolean isKeySerializerCompatibilityChecked;
 
+    private final TriConsumerWithException<
+                    Collection<IncrementalLocalKeyedStateHandle>, byte[], byte[], Exception>
+            rescalingRestoreFromLocalStateOperation;
+
```
This can be replaced by a boolean flag (`useIngestDbRestoreMode`) and a "normal" if and call in `restoreWithRescaling`.
I think it would be more readable.

rkhachatryan commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBOperationUtils.java:
```diff
@@ -123,16 +125,36 @@ public static RocksDBKeyedStateBackend.RocksDbKvStateInfo createStateInfo(
             Function<String, ColumnFamilyOptions> columnFamilyOptionsFactory,
             @Nullable RocksDbTtlCompactFiltersManager ttlCompactFiltersManager,
-            @Nullable Long writeBufferManagerCapacity) {
+            @Nullable Long writeBufferManagerCapacity,
+            List<ExportImportFilesMetaData> cfMetaDataList) {
```
Can't we use an empty list for `cfMetaDataList` instead of `null`?
If not, mark `@Nullable`?

rkhachatryan commented on flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksIncrementalCheckpointRescalingTest.java:
```diff
@@ -41,14 +41,26 @@
+        return Arrays.asList(false, true);
+    }
+
+    @Parameterized.Parameter public boolean useIngestDbRestoreMode;
```
It would be great to parameterise other unit-tests. Maybe `EmbeddedRocksDBStateBackendTest` and `FileStateBackendTest`?

StefanRRichter replied:
Not the FileStateBackend, because this only affects RocksDB. But there are also more rescaling tests that could cover this.

rkhachatryan reviewed (commented):
Thanks for rebasing the PR!
I've left some comments, PTAL.

Besides that, I think it makes sense to test the change for correctness using ITCase randomization, WDYT?
(see [`TestStreamEnvironment.randomizeConfiguration`](https://github.com/apache/flink/blob/793a66b9c6a82de9b1b0c5c547cf27d025d70f9c/flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/TestStreamEnvironment.java#L112))

StefanRRichter commented:
@rkhachatryan Can you take one more look? I have fixed all problems that I encountered in the original PR and also addressed your comments. Using `TestStreamEnvironment.randomizeConfiguration` seems a bit problematic because the RocksDB config keys are not visible in that module. We also still need to change to frocksDB version and the notice file once we have the release.

rkhachatryan commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBIncrementalCheckpointUtils.java:
```diff
@@ -154,6 +164,67 @@ private static void deleteRange(
+                        subPath.toFile()
+                                .listFiles((file, name) -> name.toLowerCase().endsWith(".sst"));
+
+                if (exportedSstFiles != null && exportedSstFiles.length > 0) {
+                    resultOutput
+                            .computeIfAbsent(stateMetaInfo, (key) -> new ArrayList<>())
+                            .add(cfMetaData);
+                }
```
`else cfMetaData.close();` ?

rkhachatryan reviewed (commented):
Thanks for updating the PR @StefanRRichter,

LGTM in general.

Regarding randomised testing,
> Using TestStreamEnvironment.randomizeConfiguration seems a bit problematic because the RocksDB config keys are not visible in that module.

Do I understand correctly, that adding rocksdb module as a test dependency would create a cycle?
In that case, should we just use string constants to enable this feature randomly?

rkhachatryan commented:
One more thing, we should probably update the UI and show the new flag under `Checkpoints / Configuration`. But that can be a separate PR.

mayuehappy commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/restore/RocksDBIncrementalRestoreOperation.java:
```diff
@@ -488,6 +495,8 @@ private void rescaleClipIngestDB(
 
+                // Check if the data in all SST files referenced in the handle is within the
+                // proclaimed key-groups range of the handle.
                 if (RocksDBIncrementalCheckpointUtils.isSstDataInKeyGroupRange(
```
Here we need to check
Is key in proclaimed range or is there overlap between the checking stateHandles?
For example, the range of proclaimed is [1,5] [6,10] but the actual range is [1,7] [8,9]. Should it be possible to export in this case?

StefanRRichter replied:
It is possible to import the case in your example, but the cost of detecting the case is that we need to compare all handles against each other which we can only do after all DBs have been opened. I'm excluding this case on purpose to avoid it, because I also assume that this missed opportunities will be very very rare. In particular because subtasks over time will compact their data to the proclaimed range from normal compaction activity.

mayuehappy replied:
haha , that makes sense to me

mayuehappy commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/restore/RocksDBIncrementalRestoreOperation.java:
```diff
@@ -539,13 +552,14 @@ private void rescaleClipIngestDB(
-            // if we have remaining handles to restore, we will insert by copy with from temporary
-            // instances to base DB.
+            // if we have remaining unopened handles to restore, we will insert by copy via
+            // temporary instances to base DB.
```
In the old code, we called `choseTheBestStateHandleForInitial` to choose the best state handle to init the initial db .Because if we use TheBestHandle as a temporary DB instead of the main DB, we may need to write a lot of data when copying. Can we maintain this logic int the new code ?
When choosing db to export, we prioritize `TheBestHandle` to ensure that we will not copy TheBestHandle during the subsequent copying phase. Can this ensure that there is no regression compared to the old code

StefanRRichter replied:
Yes, I'm already working on adding that code again here.

StefanRRichter commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/restore/RocksDBIncrementalRestoreOperation.java:
```diff
@@ -310,98 +322,275 @@ private void restoreWithRescaling(Collection<KeyedStateHandle> restoreStateHandl
+                            "Exporting state handle {} for rescaling with Clip/Ingest DB.",
                             stateHandle);
+                    // Use Range delete to clip the temp db to the target range of the backend
+                    RocksDBIncrementalCheckpointUtils.clipDBWithKeyGroupRange(
```
Makes sense to try that. Your code was doing the delete before they export, so I was also wondering about this but didn't bother to change. Let me try that.

StefanRRichter replied:
I changed the code accordingly. It didn't work at first because the import code didn't add the ColumnFamilyHandles to the DbHandle. Now it seems to work :)

pnowojski reviewed (approved):
Broadly speaking LGTM % other reviewers' open issues.
